validator/project_validator.go

```diff
@@ -1850,178 +1804,68 @@ func validateTaskSyncCommands(p *model.Project, runLong bool) ValidationErrors {
-			errMsg += " for both patches and non-patches"
-		} else if depReqs.requireOnPatches {
-			errMsg += " for patches"
-		} else if depReqs.requireOnNonPatches {
-			errMsg += " for non-patches"
-		} else if depReqs.requireOnGitTag {
-			errMsg += " for git-tag builds"
-		}
-		return errors.Errorf(errMsg, source.TaskName, source.Variant, target.TaskName, target.Variant)
-	}
-	return nil
-}
+		dependentTaskUnit := tvTaskUnitMap[model.TVPair{TaskName: dependent.Name, Variant: dependent.Variant}]
+		dependedOnTaskUnit := tvTaskUnitMap[model.TVPair{TaskName: dependedOn.Name, Variant: dependedOn.Variant}]
 
-type dependencyRequirements struct {
-	lastDepNeedsSuccess bool
-	requireOnPatches    bool
-	requireOnNonPatches bool
-	requireOnGitTag     bool
-}
+		var dep model.TaskUnitDependency
+		for _, dependency := range dependentTaskUnit.DependsOn {
+			if dependency.Name == dependedOn.Name && dependency.Variant == dependedOn.Variant {
+				dep = dependency
+			}
+		}
 
-// dependencyMustRun checks whether or not the current task in a build
-// variant depends on the success of the target task in the build variant.
-func dependencyMustRun(target model.TVPair, current model.TVPair, depReqs dependencyRequirements, allNodes []model.TVPair, visited map[model.TVPair]bool, tvToTaskUnit map[model.TVPair]model.BuildVariantTaskUnit) (bool, error) {
-	isVisited, ok := visited[current]
-	// If the node is missing, the dependency graph is malformed.
-	if !ok {
-		return false, errors.Errorf("dependency '%s' in variant '%s' is not defined", current.TaskName, current.Variant)
-	}
-	// If a node is revisited on this DFS, the dependency graph cannot be
-	// checked because it has a cycle.
-	if isVisited {
-		return false, errors.Errorf("dependency '%s' in variant '%s' is in a dependency cycle", current.TaskName, current.Variant)
-	}
+		if dependentTaskUnit.RunsOnPatches() && (!dependedOnTaskUnit.RunsOnPatches() || dep.PatchOptional) {
+			return false
+		}
+		if dependentTaskUnit.RunsOnNonPatches() && !dependedOnTaskUnit.RunsOnNonPatches() {
+			return false
+		}
+		if dependentTaskUnit.RunsOnGitTag() && !dependedOnTaskUnit.RunsOnGitTag() {
+			return false
+		}
 
-	taskUnit := tvToTaskUnit[current]
-	// Even if current depends on target according to the dependency graph, if
-	// the current task will not run in the same cases as the source (e.g. the
-	// source task runs on patches but current task does not, or if the current task
-	// is only available to git tag builds), the dependency is
-	// not reachable from this branch.
-	if depReqs.requireOnPatches && (taskUnit.SkipOnPatchBuild() || taskUnit.SkipOnNonGitTagBuild()) {
-		return false, nil
-	}
-	if depReqs.requireOnNonPatches && (taskUnit.SkipOnNonPatchBuild() || taskUnit.SkipOnNonGitTagBuild()) {
-		return false, nil
-	}
-	if depReqs.requireOnGitTag && (taskUnit.SkipOnNonPatchBuild() || taskUnit.SkipOnGitTagBuild()) {
-		return false, nil
-	}
+		if statuses != nil && dependedOn == dependedOnNode {
```

Comment:
Can you explain the dependedOn / dependedOnNode bit?
In general I think we should try explanation comments throughout, not just at the top of the function, so this is easier to read as we go in the future.

Reply:
I added a comment
```go
// If statuses is specified we need to check the edge's status when the edge points to the target node.
```